### PR TITLE
Additional Tag modifier classes for different colours

### PR DESCRIPTION
**Pull request by adamsilver**
I spoke with @36degrees and then @timpaul about contributing additional Tag component styles via modifier classes. A lot of services (citizen-facing and internal) use the Tag component for different statuses.

Most of the time, different colours are used as an enhancement to spot those different states easily. Most of the designs I've seen put white text on coloured backgrounds. At DfE we did the same but they fail contrast requirements.

So we now uses these styles which are accessible and use the tints and shades of the GOV.UK colour palette.

More information here:
https://github.com/alphagov/govuk-design-system-backlog/issues/62#issuecomment-570575887

Note:

1. I would usually look for meaningful names for modifier classes like ‘urgent’ but statuses are often unique to the system and so I used colours for the class names giving users flexibility.

2. I still need to add guidance / examples to the design system repo. I'm not sure what the process is on that i.e. do I wait for this to be merged or something more elaborate?

Screenshot:

image

### CODE REVIEW

File: src/govuk/components/tag/template.test.js [hunk 1]
```diff
@@ -75,4 +75,94 @@ describe('Tag', () => {
     expect($component.attr('data-test')).toEqual('attribute')
     expect($component.attr('id')).toEqual('my-tag')
   })
+
+  it('grey example passes accessibility tests', async () => {
```
Comment (36degrees) [inline src/govuk/components/tag/template.test.js]: The `render` function just returns the HTML from the macro itself, without the layout that brings in the stylesheet. Unfortunately this means there's little benefit to running this for every example, as it won't be checking the contrast between the colours that will actually be used.

File: src/govuk/components/tag/_tag.scss [hunk 2]
```diff
@@ -20,6 +20,8 @@
     text-decoration: none;
     text-transform: uppercase;
 
+    white-space: nowrap;
```
Comment (36degrees) [inline src/govuk/components/tag/_tag.scss]: This change seems unrelated to the commit description – if it's required, can we make this change in another commit that explains why it's being added, or possibly make it part of a separate PR?

File: src/govuk/components/tag/_tag.scss [hunk 3]
```diff
@@ -41,4 +43,50 @@
   .govuk-tag--inactive {
     background-color: govuk-colour("dark-grey", $legacy: "grey-1");
   }
+
+  .govuk-tag--grey {
```
Comment (36degrees) [inline src/govuk/components/tag/_tag.scss]: We should consider whether there's a need for `inactive` and `grey`, or whether they can be the same thing.

File: src/govuk/components/tag/tag.yaml [hunk 4]
```diff
@@ -24,3 +24,39 @@ examples:
    data:
     text: alpha
     classes: govuk-tag--inactive
+ - name: grey
```
Comment (36degrees) [inline src/govuk/components/tag/tag.yaml]: Thanks for adding these 👏

**Comment by nickcolley** [comment]
I think it's worth carefully considering the modifier names, if we name them based on their colour and decide the colours in the future need to change they may not make sense anymore or require a breaking change?

Good ol' boostrap has this sort of thing with their 'badges' and they use broad semantic names which might be worth considering:
https://getbootstrap.com/docs/4.4/components/badge/#contextual-variations

File: src/govuk/components/tag/template.test.js (hunk 1 above)
Comment (nickcolley) [reply src/govuk/components/tag/template.test.js]: We've talked about running axe on the review application itself in the future, which would achieve what you're looking to do.

Comment (adamsilver) [reply src/govuk/components/tag/template.test.js]: Shall I remove all of the tests I added @36degrees ?

File: src/govuk/components/tag/_tag.scss (hunk 3 above)
Comment (adamsilver) [reply src/govuk/components/tag/_tag.scss]: Yeah. I considered deleting it, but I thought that would delay the contribution as I assumed it would be a breaking change.

Comment (36degrees) [reply src/govuk/components/tag/_tag.scss]: We could consider marking it as deprecated, for example.

File: src/govuk/components/tag/template.test.js (hunk 1 above)
Comment (36degrees) [reply src/govuk/components/tag/template.test.js]: I think so – unfortunately we don't have a good way to test this sort of thing right now.

**Comment by adamsilver** [comment]
> I think it's worth carefully considering the modifier names, if we name them based on their colour and decide the colours in the future need to change they may not make sense anymore or require a breaking change?

@nickcolley 

Yeah—I did consider this as I said in the PR description. 

I think it would be too difficult to come up with useful names (I checked Bootstrap) and I also think it could mean code within a service could be confusing...

For example, let's say we call the orange variant ‘warning-2’ and the yellow being ‘warning-1’. (The reason for the numbers is because there aren't enough generic class names.) 

Let's imagine I use ‘warning-1’ because in Apply for teaching training we want to the ‘withdrawn’ status to be yellow. That would indicate “withdrawn’ is some sort of warning when it isn't.

File: src/govuk/components/tag/_tag.scss (hunk 3 above)
Comment (adamsilver) [reply src/govuk/components/tag/_tag.scss]: Yeah, and actually, it's not documented on the Tag component page. Is it documented publically elsewhere?

**Comment by nickcolley** [comment]
@adamsilver I think you might be right that it's too difficult to have semantics that can anticipate what context these are used in and the colours dont change often.

File: src/govuk/components/tag/_tag.scss (hunk 3 above)
Comment (adamsilver) [reply src/govuk/components/tag/_tag.scss]: I've left it in, and maybe the deprecation can be a release note? Or do I need to do something in the code to mark it as deprecated?

Comment (nickcolley) [reply src/govuk/components/tag/_tag.scss]: Since it's not been documented we could consider dropping it in 4.0.0

**Comment by timpaul** [comment]
> @adamsilver I think you might be right that it's too difficult to have semantics that can anticipate what context these are used in and the colours dont change often.

Good discussion! I'm comfortable with the colour-based class names, as they're aligned with the names in our colour palette. 

I totally agree that we can't anticipate all future semantic states that the tags will be used to represent. But we might converge on a few one day (eg. 'success', 'warning', 'fail' etc.) - does this approach block us from doing that in the future?

**Comment by adamsilver** [comment]
> But we might converge on a few one day (eg. 'success', 'warning', 'fail' etc.) - does this approach block us from doing that in the future?—@timpaul 

I think that would be fine because we could just add an additional govuk-tag--success class name as a minor release.

**Comment by peteryates** [comment]
I don't know if this has been brought up elsewhere but, to me at least, the red and orange look _very_ similar. To the point where I might not notice they were different. To a slightly lesser extent, the red and pink do too.

**Comment by adamsilver** [comment]
@peteryates some of us at DfE thought the same thing, but we weren't overly concerned because:

1. Colour is only used as an enhancement
2. You don't have to use both colours if you don't want to
3. They may not appear right next to each in the interface

**Comment by pripley-ea** [comment]
tag-colour-vision
I think these will be very useful and I like what you have done with matching the text colour with the background hue.

I've just run a quick test to see how a user with Protanopia would see them as shown above (red colour vision deficiency) and they are still usable as the text is clear but some of the colours cross over so the only thing is to be careful if multiple tags are used together. 

Really nice addition to the design system :)

File: src/govuk/components/tag/_tag.scss (hunk 3 above)
Comment (hannalaakso) [reply src/govuk/components/tag/_tag.scss]: Created an issue for dropping it in 4.0.0: https://github.com/alphagov/govuk-frontend/issues/1757  

I'll chat to @m-green about whether we want to add a deprecation notice for `govuk-tag--inactive` somewhere, seeing that we don't document the class anywhere currently.

Comment (hannalaakso) [reply src/govuk/components/tag/_tag.scss]: @adamsilver Would you be okay to add something like
```
Deprecated. We will be removing this class in a future release, use `.govuk-tag--grey` instead.
```
above the `govuk-tag--inactive` class in the scss file? 🙏

Comment (m-green) [reply src/govuk/components/tag/_tag.scss]: No probs - added in https://github.com/alphagov/govuk-frontend/pull/1756/commits/ed8534b156ad53c32dd71240b81a12b243f747a5

Comment (adamsilver) [reply src/govuk/components/tag/_tag.scss]: @hannalaakso where would you like me to put this line?

Comment (adamsilver) [reply src/govuk/components/tag/_tag.scss]: @hannalaakso done.

**Comment by hannalaakso** [comment]
@adamsilver Thanks for updating the PR. 

We're now thinking that https://github.com/alphagov/govuk-frontend/pull/1711/commits/1d79100f2db22cd6c83904624399a0c4f6b5d590 should be moved to its own PR as it could possibly be a breaking change as it could change your layout (esp on mobile) by forcing a longer tag to stay on its own line. We should talk about it separately. Apologies for the back and forth on this 🙁 

If you were able to remove the commit from this PR (let me know if you need help rebasing) we can get this one merged. Thank you!

**Comment by adamsilver** [comment]
@hannalaakso thank you—I'd be delighted to take you up on your offer to do it for me pls :)

**Comment by hannalaakso** [comment]
_On 13 February 2020 the [Design System working group approved](https://github.com/alphagov/govuk-design-system-backlog/issues/62#issuecomment-590800378) the addition of different coloured variants of this component._ 

(Just mentioning this for reference as I didn't see it explicitly mentioned in the comments on this PR.)
